### ISSUE TITLE
target-ing class methods/fields with @DexWrap/@DexEdit

### ISSUE THREAD

**Issue by Novex** [DexPatcher/dexpatcher-tool#27]
Hi! Thanks so much for this awesome tool!

I'm trying to de-obfuscate and change some proguard obfuscated code. So far I've been trying to make sense of it by renaming methods and fields using `target`. Things compile fine, but I'm getting runtime errors saying some methods/fields don't exist:

My example class:
```
// Works
@DexEdit(target = "ro.numedecod.a.e.b", contentOnly = true)
public class b_EncoderInfo {
    
    // Doesn't work
    @DexEdit(target = "g")
    public int g_initialBitrate;
    
    // Also doesn't work
    @DexWrap(target = "a")
    public String a_getDefaultEncoderType() {
        return a_getDefaultEncoderType();
    }
    
    // Works
//    @DexWrap(target = "a")
//    public String a() {
//        return a();
//    }

    // Works
//    @DexEdit(target = "a")
//    private String a_getDefaultEncoderType() { throw null; }
//
//    @DexAdd
//    public String a() {
//        return a_getDefaultEncoderType();
//    }
}
```

The class targeting works great!

The method targeting works if I manually `@DexEdit` and `@DexAdd`, but doesn't work with the `@DexWrap` helper (even though I think it's meant to be the same).

I can't get the `g_initialBitrate` field target to work at all.

I was wondering if I'm using dexpatcher properly? Or am I completely in the wrong and there's a better way to annotate obfuscated code?

**Comment by Novex** [comment]
So after looking into the DexPatcher code more I've gotten a bit more of an idea as to what can be done,

It looks like direct class substitutions are currently possible, but if that class is used in other places (eg. a method's return type/parameter) then it's not substituted there.

This is particularly problematic with obfuscated apps, where a class (eg. `app.a`) having the same name as a package prevents the compiler being able to reference any of the package classes (eg. `app.a.b` can't be referenced).

Direct method substitutions are possible too, unless they have a class parameter "obscured" by the package/class conflict. The java compiler won't let you reference the original class and DexPatcher doesn't rewrite method parameters/return types that target another class.

It's currently not possible to use targeted references to field/methods either since DexPatcher currently doesn't rewrite method implementations in that way (though it looks like the groundwork is there).

I think it would be useful to have the ability to do all this (and would definitely help at least me in being able to make sense of obfuscated apps).

If you think it's a good idea, my plan is to:

Pre-process the patch file:
* A first time to get a map of patch classes to target classes
  * eg. `ro.numedecod.a.e.b_EncoderInfo => ro.numedecod.a.e.b`
* And again a second time to:
  * For each field, replace its type class with the target class and make a map of patched field to target field
    * eg. `ro.numedecod.a.e.b.g_InitialBitrate => ro.numedecod.a.e.b.g`
  * For each method, replace its return type/parameter classes with their target classes and make a map of patched method to target method
    * eg. `ro.numedecod.a.e.b.a_getDefaultEncoderType() => ro.numedecod.a.e.b.a()`

Then during the main DexPatcher processing:
* For each field, replace its type class with the target class
* For each method, replace its return type/parameter classes with their target classes
  * For each method implementation:
    * Use the pre-processed method map to replace the dangling patched method invocation signatures with their targets
    * Use the pre-processed field map to replace the dangling patched field signatures with their targets

Hopefully the end result is that patch classes can use any targeted patch class/method/field and DexPatcher will translate them to their target references automatically.

**Comment by Lanchon** [comment]
hey! thanks for using dexpatcher. im very sorry for the delay, im travelling at the moment. i should be able to reply in about a week. but you kinda got it: explicit targeting is used to rename items, such as methods or whole classes, while at the same time declaring the newly named item for use within the patch. it is not a mechanism to handle obfuscated classes. the case of contentOnly=true is atypical in dxp and has some use regarding obfuscation, but it is not a general solution. more to come later, have to go now, sorry.

**Comment by Novex** [comment]
No worries at all - appreciate you taking the time to reply and enjoy your travels!

**Comment by Lanchon** [comment]
hi,

the dxp patch language as implemented today does not support obfuscation.

well, actually it sort of does. the real issue is that it does not attempt to cure the naming limitations of the java language w.r.t. JVM bytecode.

supporting obfuscation is more complex than these simple naming issues, and entails varied stuff:

1. support decompilation with iterative renaming to aid in code analysis.
2. support the patching of different versions of the obfuscated APK (ie, different runs of proguard) with the same patch definition.
3. continue to support on-device patching.
4. allow using self-documenting readable names in patch code.
5. provide a workaround for java language naming limitations. (your concern!)
6. etc...

i've given a lot of though to the subject, and designed The Grand Scheme Of Things (tm) to support patching of obfuscation code.

unfortunately, i also realized that i simply don't have the time needed to implement this grand scheme. so this scheme is flawed beyond repair in that it would never work because it would never be finished.

but something *is* needed... so i also designed a much more economical solution, simplified and reduced in cost, but achievable. this solution at least meets bullets 3, 4 and 5, and can be implemented quickly, and will be at some point.

next in line are two missing features of dexpatcher-tool: this obfuscation support and the `@DexCheck` tag. the new tag is much more expensive in development time, but for better or worse, i chose it would be the next feature because i'm so ashamed by the absence of this tag. had i chosen otherwise, obfuscation support would be published already.

so my message is:
- maybe not fork dxp: the feature is coming anyway.
- don't try to use dxp as is for obfuscated code; it won't work.
- but you can try a workaround.

workarounds:
- there are some jar class renaming tools. if you can integrate one into the build system you can rename, say, `org.whatever.okhttp.HttpRequest` into `a.a.a.a` right after patch compilation, according to a configuration file. look for a tool that somebody already integrated into the android build system if possible. you would need to set `importSymbols = false` in dxp.
- use a JVM-backed language that does not have the naming shortcomings of Java. possible examples include Kotlin and Scala, both of which i believe support quoting for werid names.

so there, a short intro to the problem. i'll implement as time permits.

**Comment by Novex** [comment]
Hmm.. I think I managed to implement class renaming in dexpatcher itself if that's something you'd be interested in adding? It seems to take care of goals 4 and 5 (possibly also 3 as it doesn't introduce any new dependencies) and may serve as a stop gap until the perfect solution gets done (which sounds amazing).

I'm not sure if I implemented it particularly well, but I'll open a pull request and it might be faster for you to fix it or provide feedback than start from scratch. I'm still working through my reversing project and finding edge cases, though it seems to work pretty well so far.

I appreciate the workarounds as well - I didn't even think about using Kotlin to get around the naming collisions!

**Comment by Novex** [comment]
Hmm.. I think I managed to implement class renaming in dexpatcher itself if that's a contribution you'd be interested in accepting? It seems to take care of goals 4, 5 and possibly 3 (as it doesn't introduce any new dependencies) and may serve as a stop gap until the perfect solution is complete (which sounds amazing).

I'm not sure if I implemented it particularly well, but I'll open a pull request and it might be faster for you to fix it or offer feedback than design from scratch. I'm still working through my reversing project and finding edge cases, though it seems to be working pretty well so far.

I appreciate the workarounds as well - I didn't even think about using Kotlin to get around the naming collisions!
